### PR TITLE
[pipelining] Fix more leaks and check leaks in tests

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136678
* __->__ #136584
* #136507

Fix two more leaks of the same variety as #136507 (see that PR desc and attached gdoc for debug details).

This time, also add a test-time check that helped to discover new leaks and ensure we won't accidently regress.

Adds `check_tensor_leak` util which internally asserts no tensors are being kept alive by other objects involved in py ref cycles.

Uses objgraph for a nice debug utility when a leak is found.

Credit to @H-Huang for pointing out objdump and helping debug the 'param_group["intermediates"]` leak.

I manually confirmed that all 3 of the leaks identified/fixed so far are caught by the unit test and checker.

Sample output, if I re-introduce a leak by commenting out `del param_group["intermediates"]` in _backward.py,
and run `python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`:

```
warnings.warn(
/data/users/whc/pytorch/torch/testing/_internal/common_utils.py:5341: UserWarning: 34 tensors were found in the garbage. Did you introduce a reference cycle?
warnings.warn(
/data/users/whc/pytorch/torch/testing/_internal/common_utils.py:5347: UserWarning: Dumping first 1 objgraphs of leaked tensors rendered to png
Graph written to /tmp/objgraph-ztz642h3.dot (19 nodes)
Graph viewer (xdot) not found, generating a png instead
Image generated as /tmp/objgraph-ztz642h3.png
```

rendering of ` /tmp/objgraph-ztz642h3.png`:
image

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -5316,3 +5316,46 @@ def repl_frame(m):
...
+    try:
+        gc.collect()
+        gc.set_debug(gc.DEBUG_SAVEALL)
+
+        # run the user code, after cleaning any existing refcycles
+        yield
+
+        gc.collect()
+        garbage_tensors = sum(int(isinstance(g, torch.Tensor)) for g in gc.garbage)
+        if garbage_tensors > 0:
```
Comment (wconstab) [inline torch/testing/_internal/common_utils.py:5340]: oops, when i refactored this to a util, i dropped the `self.assertFalse` part so the tests won't actually fail anymore.  need to fix that...

**Review by kwen2501 — APPROVED** [review]
Thanks for the great tool!

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -5316,3 +5316,46 @@ def repl_frame(m):
...
+    since this is likely a more serious leak than typical python refcycles.
+
+    limit specifies how many tensors to dump debug graphs for (default=1)
+    """
+    try:
+        gc.collect()
+        gc.set_debug(gc.DEBUG_SAVEALL)
+
+        # run the user code, after cleaning any existing refcycles
+        yield
```
Comment (H-Huang) [inline torch/testing/_internal/common_utils.py:5336]: For testing mentioned below or for better observability you could yield a `Result` class or something similar that has attributes like `has_leak` or `leaked_tensors` and then in the context manager use it like

```python
with check_leaked_tensors() as result:
   ... code
assert result.has_leak == False
```

**Review by H-Huang — APPROVED** [review]
Great addition!

File: torch/testing/_internal/common_utils.py (hunk 2 above)
Comment (wconstab) [reply torch/testing/_internal/common_utils.py]: oh, i worked on the fix before i saw your comment. i ended up just yielding the garbage_objs list directly.

File: torch/distributed/pipelining/_backward.py [hunk 3]
```diff
@@ -240,6 +240,13 @@ def stage_backward_weight(
         )
         weights_edges = tuple(GradientEdge(w, 0) for w in param_group["params"])
 
+        # Break a reference cycle caused inside stage_backward_input->get_hook->hook
+        # The summarized cycle is:
+        # `hook` -> cell -> param_group -> intermediates -> `hook`
+        # becuase we install the hook function onto each of the intermediate autograd nodes.
+        # We need to keep intermediates alive up until backward_weight, but we can free it now.
+        # del param_group["intermediates"]
```
Comment (H-Huang) [inline torch/distributed/pipelining/_backward.py:248]: accidentally commented this?

Comment (wconstab) [reply torch/distributed/pipelining/_backward.py:248]: Thanks! fixed.

**Pull request closed without merging**